src/rayon_impls.rs

```diff
@@ -0,0 +1,117 @@
+use crate::HashMap;
+use rayon::iter::{FromParallelIterator, IntoParallelIterator, ParallelExtend, ParallelIterator};
+use std::hash::{BuildHasher, Hash};
+
+impl<K, V, S> ParallelExtend<(K, V)> for HashMap<K, V, S>
```

Comment:
Could we implement this for `HashMapRef` as well?

Reply:
Not really, since its `Guard` is stuck on one thread.

I guess you could hack it by ignoring the ref's own guard and extending the map inside.

Reply:
Yeah, that was basically what I was thinking. Not sure how much that _really_ buys us, but if we're trying to encourage users to use `HashMapRef` over dealing with guards themselves, it might be a nice ergonomic improvement.

Reply:
FWIW, `HashMapRef` doesn't even have plain `Extend` yet. :wink: 

The `HashMap` implementations don't make you deal with guards anyway, since the trait methods don't really make that possible, but it seems fine to keep API parity and add that to `HashMapRef` too.

While we're talking parity, there's only `Extend<(K, V)> for &HashMap`, not `for HashMap`, but this PR has `ParallelExtend` for both. But there is also `Extend<(&K, &V)> for &HashMap` with `Copy` types, which this PR doesn't have.

Then multiply all of this for sets...

Reply:
Bah... I'm honestly not sure how best to deal with that. I'm inclined to remove the `where Copy` impl. The impls for `&HashMap` I think make sense for us (arguably more so than "regular" `Extend`). I think we could go as far as to remove the `impl for HashMap` entirely, and _only_ implement for `&`.

Reply:
Iterator methods like `partition` and `unzip` need `Extend` for the collection itself. If you only implement for `&`, then it's basically only useful for direct `extend(...)` calls.

That matters less for `HashMapRef`, which is already a reference of sorts. I would just do that one directly, not `for &HashMapRef`, but it can't be a target of `unzip`/etc. anyway.

`Copy` impls are debatable, but they're at least consistent with `std` collections.

Reply:
Oh, interesting, I didn't even know that. I suppose they all sort of need to stay then (well, `Copy` I think is still unclear, given we now have `Iterator::copied`). I agree with you that impls for `&HashMapRef` are probably unnecessary.

Reply:
Well, `copied()` doesn't work for `Item=(&K, &V)` since that tuple itself is not a reference.

Reply:
The heat in Boston is clearly getting to me. And that explains why the standard library `HashMap` has an implementation for `Copy` (and why we probably should too).